### PR TITLE
veleroDeployment: Create a Service and ServiceMonitor

### PR DESCRIPTION
**Pull request by mbarnes**
Create Service and ServiceMonitor so that metrics can be scraped

### CODE REVIEW

File: deploy/role.yaml [hunk 1]
```diff
@@ -32,8 +32,7 @@ rules:
   resources:
   - servicemonitors
   verbs:
-  - get
-  - create
+  - '*'
```
Comment (cblecker) [inline deploy/role.yaml]: Per https://github.com/operator-framework/operator-sdk/issues/671, can we swap out the * for:
```suggestion
  - create
  - delete
  - get
  - list
  - patch
  - update
  - watch
```

File: pkg/controller/velero/velero.go [hunk 2]
```diff
@@ -158,6 +162,60 @@ func (r *ReconcileVelero) provisionVelero(reqLogger logr.Logger, namespace strin
 		}
 	}
 
+	// Install Service
```
Comment (cblecker) [inline pkg/controller/velero/velero.go]:
```suggestion
	// Install Metrics Service
```

File: pkg/controller/velero/velero.go [hunk 3]
```diff
@@ -294,6 +352,35 @@ func veleroDeployment(namespace string, veleroImage string) *appsv1.Deployment {
 	return deployment
 }
 
+func veleroService(namespace string) *corev1.Service {
```
Comment (cblecker) [inline pkg/controller/velero/velero.go]: So I'd like to future-proof this function a bit with a bit more autogeneration and less hard coding of values.

My suggestion:
we create a function that is `func createServiceFromDeployment(d *appsv1.Deployment) *corev1.Service`

Then based on the deployment, we can generate most of the details:
- Service name := `d.ObjectMeta.Name` + "-metrics"
- Namespace should match the deployment
- Loop through all containers and all the ports on the containers, and anyone that has "metrics" in the name should be added to the ServicePorts
- Selector should match the deployment.Spec.Template labelset

File: pkg/controller/velero/velero.go [hunk 4]
```diff
@@ -158,6 +162,60 @@ func (r *ReconcileVelero) provisionVelero(reqLogger logr.Logger, namespace strin
...
+	if !reflect.DeepEqual(foundService.Spec, service.Spec) {
+		// Specs aren't equal, update and fix.
+		reqLogger.Info("Updating Service", "foundService.Spec", foundService.Spec, "service.Spec", service.Spec)
+		foundService.Spec = *service.Spec.DeepCopy()
+		if err = r.client.Update(context.TODO(), foundService); err != nil {
+			return reconcile.Result{}, err
+		}
+	}
+
+	// Install ServiceMonitor
```
Comment (cblecker) [inline pkg/controller/velero/velero.go]:
```suggestion
	// Install Metrics ServiceMonitor
```

File: pkg/controller/velero/velero.go (hunk 3 above)
Comment (mbarnes) [reply pkg/controller/velero/velero.go]: :+1: I like it.

**Comment by mbarnes** [comment]
Rebased on the latest `master` branch revision and squashed in all suggestions except for the service refactoring.

File: pkg/controller/velero/velero.go [hunk 5]
```diff
@@ -297,6 +361,48 @@ func veleroDeployment(namespace string, veleroImageRegistry string) *appsv1.Depl
...
+
+	return &corev1.Service{
+		TypeMeta: metav1.TypeMeta{
+			Kind:       "Service",
+			APIVersion: corev1.SchemeGroupVersion.String(),
+		},
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      deployment.ObjectMeta.Name + "-metrics",
+			Namespace: deployment.ObjectMeta.Namespace,
+			Labels:    map[string]string{"name": "velero"},
```
Comment (cblecker) [inline pkg/controller/velero/velero.go]:
```suggestion
			Labels:    map[string]string{"name": deployment.ObjectMeta.Name},
```

File: pkg/controller/velero/velero.go [hunk 6]
```diff
@@ -297,6 +361,48 @@ func veleroDeployment(namespace string, veleroImageRegistry string) *appsv1.Depl
 	return deployment
 }
 
+func veleroServiceFromDeployment(deployment *appsv1.Deployment) *corev1.Service {
```
Comment (cblecker) [inline pkg/controller/velero/velero.go]:
```suggestion
func merticsServiceFromDeployment(deployment *appsv1.Deployment) *corev1.Service {
```

File: pkg/controller/velero/velero.go [hunk 7]
```diff
@@ -158,6 +160,68 @@ func (r *ReconcileVelero) provisionVelero(reqLogger logr.Logger, namespace strin
 		}
 	}
 
+	// Install Metrics Service
+	foundService := &corev1.Service{}
+	service := veleroServiceFromDeployment(deployment)
```
Comment (cblecker) [inline pkg/controller/velero/velero.go]:
```suggestion
	service := merticsServiceFromDeployment(deployment)
```

**Comment by mbarnes** [comment]
Reapplying  **lgtm** since I just squashed @cblecker's suggestions into the 2nd commit.  No other changes.

**Pull request merged**